### ISSUE TITLE
Compilation errors when generating a mutation with NonNullable Input

### ISSUE THREAD

**Issue by backjo** [apollographql/apollo-codegen#185]
When doing the following mutation in Swift, the generated classes do not compile

Schema:
```
type Author {
    id: Int!
    firstName: String
    lastName: String
    posts: [Post] # the list of Posts by this author
  }

  input AuthorInput {
    id: Int!
    firstName: String
	  lastName: String
		posts: PostInput
  }

  type Post {
    id: Int!
    title: String
    author: Author
    votes: Int
  }

  input PostInput {
    id: Int!
		firstName: String
		lastName: String
	  author: AuthorInput
  }

  # the schema allows the following query:
  type Query {
    posts: [Post]
    author(id: Int!): Author
  }

  # this schema allows the following mutation:
  type Mutation {
    upvotePost (
      postId: Int!
    ): Post
		
		createPost (
		   input: PostInput!
		) : Post
  }
```

Mutation 
```
mutation x($input:PostInput!) {
    createPost(input:$input)
}
```

Errors:
```
Value type 'AuthorInput' cannot have a stored property that references itself
Value type 'PostInput' cannot have a stored property that references itself
```

**Comment by pranayairan** [comment]
@martijnwalraven is apollo android uses same code gen? we see the issue there as well.

**Comment by martijnwalraven** [comment]
@pranayairan: Apollo Android uses `apollo-codegen` to generate an intermediate representation, but the code generation itself is written in Java. It may very well have the same issue though.

**Comment by weapon3042** [comment]
@martijnwalraven There are at least 3 solutions I found:

Solution 1: Change at least one of the structs to class, to make a reference cycle, you need at least one reference.

Solution 2: As mentioned here: [indirect enums and structs](https://stackoverflow.com/questions/37533006/indirect-enums-and-structs), an elegant solution could be to wrap your reference inside an enum with an indirect case.

In you example, it would give this:

enum ReferencePin {
  case none
  indirect case pin(Pin)
}

struct DetailedPin {
     var pin: ReferencePin
}

struct Pin {
    var detailedPin: DetailedPin?
}
or directly, depending on your needs:

enum ReferencePin {
  case none
  indirect case pin(Pin)
}

struct Pin {
    var otherPin: ReferencePin
}
Indirect tells the compiler there is a recursion in your enum, and it should insert the necessary layer of indirection.

Solution 3: 
//  Provides indirection for a given instance.
//  For value types, value semantics are preserved.
struct Indirect<T> {
    // Class wrapper to provide the actual indirection.
    private final class Wrapper {
        var value: T
        init(_ value: T) {
            self.value = value
        }
    }

    private var wrapper: Wrapper

    init(_ value: T) {
        wrapper = Wrapper(value)
    }

    var value: T {
        get {
            return wrapper.value
        }
        set {
            // upon mutation of value, if the wrapper class instance is unique,
            // simply mutate the underlying value directly.
            // otherwise, create a new instance.
            if isKnownUniquelyReferenced(&wrapper) {
                wrapper.value = newValue
            } else {
                wrapper = Wrapper(newValue)
            }
        }
    }
}

You can now just use the Indirect wrapper for one (or both) of your structs properties:

struct DetailedPin {
    var pin: Indirect<Pin>?
}

struct Pin {
    var foo: String
    var detailedPin: DetailedPin?
}


var p = Pin(foo: "foo", detailedPin: nil)
var d = DetailedPin(pin: Indirect(p))
p.detailedPin = d

**Comment by martijnwalraven** [comment]
This should be fixed in apollo-codegen 0.16.2.

@weapon3042: Thanks for looking into solutions for avoiding reference cycles with value types. In this case, the easiest solution seemed to be to replace stored properties with computed properties, because the end result should be a `graphQLMap` anyway.

**Comment by weapon3042** [comment]
@martijnwalraven Thanks!
